database/src/service/read.rs

```diff
@@ -228,57 +317,191 @@ fn map_request(
+    let (tx_ro, tables) = set_tx_ro_and_tables!(env, env_inner);
+
+    // -> Result<(AmountIndex, OutputOnChain), RuntimeError>
+    let inner_map = |amount, amount_index| {
+        let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+
+        let pre_rct_output_id = PreRctOutputId {
+            amount,
+            amount_index,
+        };
+        let output = get_output(&pre_rct_output_id, tables.outputs())?;
+
+        //---  Map `Output` -> `OutputOnChain`
+
+        // FIXME: implement lookup table for common values:
+        // <https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/ringct/rctOps.cpp#L322>
+        let commitment = ED25519_BASEPOINT_POINT + H() * Scalar::from(amount);
+
+        let time_lock = if output
+            .output_flags
+            .contains(OutputFlags::NON_ZERO_UNLOCK_TIME)
+        {
+            // TODO: how to recover the timelock height/time?
+            todo!()
+        } else {
+            Timelock::None
+        };
+
+        let key = CompressedEdwardsY::from_slice(&output.key)
+            .map(|y| y.decompress())
+            .unwrap_or(None);
+
+        let output_on_chain = OutputOnChain {
+            height: u64::from(output.height),
+            time_lock,
+            key,
+            commitment,
+        };
+
+        Ok((amount_index, output_on_chain))
+    };
+
+    let map = map
+        .into_par_iter()
+        .map(|(amount, amount_index_set)| {
+            Ok((
+                amount,
+                amount_index_set
+                    .into_par_iter()
+                    .map(|amount_index| inner_map(amount, amount_index))
+                    .collect::<Result<HashMap<AmountIndex, OutputOnChain>, RuntimeError>>()?,
+            ))
+        })
+        .collect::<Result<HashMap<Amount, HashMap<AmountIndex, OutputOnChain>>, RuntimeError>>()?;
+
+    Ok(Response::Outputs(map))
 }
 
 /// [`ReadRequest::NumberOutputsWithAmount`].
-/// TODO
 #[inline]
-#[allow(clippy::needless_pass_by_value)] // TODO: remove me
-fn number_outputs_with_amount(env: &Arc<ConcreteEnv>, vec: Vec<Amount>) -> ResponseResult {
-    todo!()
+fn number_outputs_with_amount(env: &ConcreteEnv, amounts: Vec<Amount>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let (tx_ro, tables) = set_tx_ro_and_tables!(env, env_inner);
+
+    let map = amounts
+        .into_par_iter()
+        .map(|amount| {
+            let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+
+            match tables.num_outputs().get(&amount) {
+                // INVARIANT: #[cfg] @ lib.rs asserts `usize == u64`
+                #[allow(clippy::cast_possible_truncation)]
+                Ok(count) => Ok((amount, count as usize)),
+                Err(e) => Err(e),
+            }
+        })
+        .collect::<Result<HashMap<Amount, usize>, RuntimeError>>()?;
+
+    Ok(Response::NumberOutputsWithAmount(map))
 }
```

Comment:
What is the valid response to `ReadResponse::NumberOutputsWithAmounts` where the `amount` doesn't exist, i.e. there are 0 outputs with `amount`.

Should it be `Err(KeyNotFound)` or `Ok(0)`?

Reply:
It should be `Ok(0)`